🔥🚀 feat: CDN (Firebase) & feat: account section

berry-13 opened the pull request:
## Summary

I have implemented CDN support using Firebase. Currently, it's specifically designed for uploading profile pictures, but we plan to extend its functionality to support images within the chat and file uploads as well. Additionally, there have been some minor graphical changes, including new icons, and a reorganization of folders in the Settings Tabs. The CDN for profile pictures is currently working for Google, Facebook, Github, and Discord (OpenID support is not yet implemented). Before being uploaded, the images undergo transformation and conversion into the webp format. On average, a profile picture weighs around 3kb, with some as low as 300B. The upload profile picture functionality works for those who don't use Firebase as well. Also, DALL-E plugin support firebase


Here are some frontend images showcasing the image upload

![image](https://github.com/danny-avila/LibreChat/assets/81851188/376a6492-450d-45b0-88d6-c423bd4166e1)

![image](https://github.com/danny-avila/LibreChat/assets/81851188/5a193f08-e291-44b9-b58f-bbe66233b867)

https://github.com/danny-avila/LibreChat/assets/81851188/d1ed5973-1546-4101-a112-bbb1715aef57

### fixed some bugs in the settings tab when the theme was light

![image](https://github.com/danny-avila/LibreChat/assets/81851188/41967057-cd97-41f7-b13f-ad6231f1d6e0)

## Change Type

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Style update
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update 

## Testing
...

- tested automatic disable "auto avatar upload with social logins":  Google, Facebook, Github and Discord
- refactor login tested with Google, Facebook, Github and Discord

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented in any complex areas of my code
- [x] I have made pertinent documentation changes
- [x] My changes do not introduce new warnings
- [x] I have written tests demonstrating that my changes are effective or that my feature works
- [x] Local unit tests pass with my changes
- [x] Any changes dependent on mine have been merged and published in downstream modules.

berry-13 committed:
localization + api-endpoint

berry-13 committed:
docs: added firebase documentation

berry-13 committed:
chore: icons

berry-13 committed:
chore: SettingsTabs

berry-13 committed:
feat: account pannel; fix: gear icons

berry-13 committed:
docs: position update

berry-13 committed:
feat: firebase

berry-13 committed:
feat: plugin support

berry-13 committed:
route

berry-13 committed:
fixed bugs with firebase and moved a lot of files

berry-13 committed:
chore(DALLE3): using UUID v4

berry-13 committed:
feat: support for social strategies; moved '/images' path

berry-13 committed:
fix: data ignored

berry-13 committed:
gitignore update

Berry-13 commented on .gitignore:
```diff
@@ -82,4 +82,6 @@ data.ms/*
\ No newline at end of file
+/images
+
+!client/src/components/Nav/SettingsTabs/Data/
```
I have reorganized all the settings tabs, but it seems that "data/" is being ignored. Therefore, I added `!client/src/components/Nav/SettingsTabs/Data/` to include `Data.tsx` in commits

danny-avila commented:
Did you create new backend tests? I believe you need to import the mocked logger, there are some tests that do this already not sure which.

> TypeError: path.basename is not a function

berry-13 committed:
Merge branch 'main' into firebase-cdn-new

danny-avila commented:
> Did you create new backend tests? I believe you need to import the mocked logger, there are some tests that do this already not sure which.
> 
> > TypeError: path.basename is not a function

@Berry-13 I would like to merge this soon so I will make edits from here if you are mostly done.

Berry-13 commented:
> > Did you create new backend tests? I believe you need to import the mocked logger, there are some tests that do this already not sure which.
> > 
> > > TypeError: path.basename is not a function
> 
> @Berry-13 I would like to merge this soon so I will make edits from here if you are mostly done.

sure, thank you!

danny-avila committed:
docs: update firebase guide

danny-avila committed:
refactor: Firebase
- use singleton pattern for firebase initialization, initially on server start
- reorganize imports, move firebase specific files to own service under Files
- rename modules to remove 'avatar' redundancy
- fix imports based on changes

danny-avila committed:
ci(DALLE/DALLE3): fix tests to use logger and new expected outputs, add firebase tests

danny-avila committed:
refactor(loadToolWithAuth): pass userId to tool as field

danny-avila committed:
feat(images/parse): feat: Add URL Image Basename Extraction

Implement a new module to extract the basename of an image from a given URL. This addition includes the  function, which parses the URL and retrieves the basename using the Node.js 'url' and 'path' modules. The function is documented with JSDoc comments for better maintainability and understanding. This feature enhances the application's ability to handle and process image URLs efficiently.

danny-avila committed:
refactor(addImages): function to use a more specific regular expression for observedImagePath based on the generated image markdown standard across the app

danny-avila committed:
refactor(DALLE/DALLE3): utilize `getImageBasename` and `this.userId`; fix: pass correct image path to firebase url helper

danny-avila committed:
fix(addImages): make more general to match any image markdown descriptor

danny-avila committed:
fix(parse/getImageBasename): test result of this function for an actual image basename

danny-avila committed:
ci(DALLE3): mock getImageBasename

danny-avila commented:
My /api review is done, I just want to fix a few things on the frontend now

danny-avila commented:
Noteable changes in the backend outside of the main firebase update:

- DALL-E/DALL-E-3 now uses the image basename from a url to test whether the DALL-E did correctly generate an image.

This fixes an issue where a reverse proxy may generate a different image url like so: media.mandrillai.tech/herearerandomcharscters.jpg

- When tools are loaded, they will pass the userId in case they are needed, such as the case with DALL-E

@Berry-13 following are some todos, which are not needed to merge the PR but will be needed some time after

- TODO: do not use Model directly, such as when you use `User.findById`, instead use a service method that uses the model; for example, we need a method similar to `getMessages` but instead `getUserById`
     - this is done incorrectly a few other places in the codebase (not by you)
- TODO: JSDocs are needed for all new functions. Should be typed out as much as possible.
- TODO: stable diffusion needs the same treatment as the DALL-E plugin

danny-avila commented on client/src/components/Nav/SettingsTabs/Account/Avatar.tsx:
```diff
@@ -0,0 +1,152 @@
+      formData.append('input', input, input.name);
+      formData.append('manual', 'true');
+
+      const response = await fetch('/api/files/images/avatar', {
```
we do not want to make raw fetch calls, I'll be replacing this with react-query

danny-avila committed:
refactor(AuthContext): use Recoil atom state for user

danny-avila committed:
feat: useUploadAvatarMutation, react-query hook for avatar upload

danny-avila committed:
fix(Toast): stack z-order of Toast over all components (1000)

danny-avila committed:
refactor(showToast): add optional status field to avoid importing NotificationSeverity on each use of the function

danny-avila committed:
refactor(routes/avatar): remove unnecessary get route, get userId from req.user.id, require auth on POST request

danny-avila committed:
chore(uploadAvatar): TODO: remove direct use of Model, `User`

danny-avila committed:
fix(client): fix Spinner imports

danny-avila committed:
refactor(Avatar): use react-query hook, Toast, remove unnecessary states, add optimistic UI to upload

danny-avila committed:
fix(avatar/localStrategy): correctly save local profile picture and cache bust for immediate rendering; fix: firebase init info message (only show once)

danny-avila committed:
fix: use `includes` instead of `endsWith` for checking manual query of avatar image path in case more queries are appended (as is done in avatar/localStrategy)

danny-avila commented:
Noteable changes in the frontend outside of the main firebase/profile picture update:

- User object now uses recoil atom state instead of useState.

Allows changing the avatar URL to dynamically render across the app once uploaded. I've observed no issues with this change as AuthContext still handles the user object the same.

- Toast has a new z-index order of 1000 to show on top of dialog overlays
